// @pothos/core/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get ArgumentRef () {
        return ArgumentRef;
    },
    get BaseTypeRef () {
        return BaseTypeRef;
    },
    get BuildCache () {
        return _buildcache.BuildCache;
    },
    get BuiltinScalarRef () {
        return _builtinscalar.BuiltinScalarRef;
    },
    get EnumRef () {
        return EnumRef;
    },
    get FieldBuilder () {
        return FieldBuilder;
    },
    get FieldRef () {
        return FieldRef;
    },
    get ImplementableInputObjectRef () {
        return ImplementableInputObjectRef;
    },
    get ImplementableInterfaceRef () {
        return ImplementableInterfaceRef;
    },
    get ImplementableObjectRef () {
        return ImplementableObjectRef;
    },
    get InputFieldBuilder () {
        return InputFieldBuilder;
    },
    get InputFieldRef () {
        return InputFieldRef;
    },
    get InputListRef () {
        return InputListRef;
    },
    get InputObjectRef () {
        return InputObjectRef;
    },
    get InputTypeRef () {
        return _input1.InputTypeRef;
    },
    get InterfaceFieldBuilder () {
        return InterfaceFieldBuilder;
    },
    get InterfaceRef () {
        return InterfaceRef;
    },
    get ListRef () {
        return ListRef;
    },
    get MutationFieldBuilder () {
        return MutationFieldBuilder;
    },
    get MutationRef () {
        return _mutation1.MutationRef;
    },
    get ObjectFieldBuilder () {
        return ObjectFieldBuilder;
    },
    get ObjectRef () {
        return ObjectRef;
    },
    get OutputTypeRef () {
        return _output.OutputTypeRef;
    },
    get QueryFieldBuilder () {
        return QueryFieldBuilder;
    },
    get QueryRef () {
        return _query1.QueryRef;
    },
    get RootFieldBuilder () {
        return RootFieldBuilder;
    },
    get ScalarRef () {
        return ScalarRef;
    },
    get SubscriptionFieldBuilder () {
        return SubscriptionFieldBuilder;
    },
    get SubscriptionRef () {
        return _subscription1.SubscriptionRef;
    },
    get UnionRef () {
        return UnionRef;
    },
    get default () {
        return _default;
    }
});
const _builder = require("./builder");
const _builder1 = require("./fieldUtils/builder");
const _input = require("./fieldUtils/input");
const _interface = require("./fieldUtils/interface");
const _mutation = require("./fieldUtils/mutation");
const _object = require("./fieldUtils/object");
const _query = require("./fieldUtils/query");
const _root = require("./fieldUtils/root");
const _subscription = require("./fieldUtils/subscription");
const _arg = require("./refs/arg");
const _base = require("./refs/base");
const _enum = require("./refs/enum");
const _field = require("./refs/field");
const _inputfield = require("./refs/input-field");
const _inputlist = require("./refs/input-list");
const _inputobject = require("./refs/input-object");
const _interface1 = require("./refs/interface");
const _list = require("./refs/list");
const _object1 = require("./refs/object");
const _scalar = require("./refs/scalar");
const _union = require("./refs/union");
require("./types/global");
_export_star(require("./errors"), exports);
_export_star(require("./plugins"), exports);
_export_star(require("./types"), exports);
_export_star(require("./utils"), exports);
const _buildcache = require("./build-cache");
const _builtinscalar = require("./refs/builtin-scalar");
const _input1 = require("./refs/input");
const _mutation1 = require("./refs/mutation");
const _output = require("./refs/output");
const _query1 = require("./refs/query");
const _subscription1 = require("./refs/subscription");
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
const SchemaBuilder = _builder.SchemaBuilder;
const _default = SchemaBuilder;
const FieldBuilder = _builder1.FieldBuilder;
const RootFieldBuilder = _root.RootFieldBuilder;
const QueryFieldBuilder = _query.QueryFieldBuilder;
const MutationFieldBuilder = _mutation.MutationFieldBuilder;
const SubscriptionFieldBuilder = _subscription.SubscriptionFieldBuilder;
const ObjectFieldBuilder = _object.ObjectFieldBuilder;
const InterfaceFieldBuilder = _interface.InterfaceFieldBuilder;
const InputFieldBuilder = _input.InputFieldBuilder;
const BaseTypeRef = _base.BaseTypeRef;
const EnumRef = _enum.EnumRef;
const InputObjectRef = _inputobject.InputObjectRef;
const ImplementableInputObjectRef = _inputobject.ImplementableInputObjectRef;
const InputListRef = _inputlist.InputListRef;
const InterfaceRef = _interface1.InterfaceRef;
const ImplementableInterfaceRef = _interface1.ImplementableInterfaceRef;
const ObjectRef = _object1.ObjectRef;
const ImplementableObjectRef = _object1.ImplementableObjectRef;
const ScalarRef = _scalar.ScalarRef;
const UnionRef = _union.UnionRef;
const ListRef = _list.ListRef;
const FieldRef = _field.FieldRef;
const InputFieldRef = _inputfield.InputFieldRef;
const ArgumentRef = _arg.ArgumentRef;
