// @pothos/plugin-prisma/lib/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
const _nodefs = require("node:fs");
const _nodepath = require("node:path");
const _generatorhelper = require("@prisma/generator-helper");
const _typescript = /*#__PURE__*/ _interop_require_wildcard(require("typescript"));
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const MIN_TS_VERSION = [
    4,
    5,
    2
];
const versionParts = _typescript.version.split(/[.-]/g).map((part)=>Number.parseInt(part, 10));
const modifiersArg = versionParts[0] >= 5 ? [] : [
    []
];
function checkVersion(expected) {
    for(let i = 0; i < 3; i += 1){
        const part = versionParts[i];
        if (part < expected[i]) {
            return false;
        }
        if (part > expected[i]) {
            return true;
        }
    }
    return true;
}
const defaultOutput = (0, _nodepath.resolve)(__dirname, '../generated.d.ts');
(0, _generatorhelper.generatorHandler)({
    onManifest: ()=>({
            prettyName: 'Pothos integration',
            defaultOutput
        }),
    onGenerate: async (options)=>{
        var _options_otherGenerators_find_output, _options_otherGenerators_find, _options_generator_output;
        if (!checkVersion(MIN_TS_VERSION)) {
            throw new Error(`@pothos/plugin-prisma requires typescript version >${MIN_TS_VERSION.join('.')}`);
        }
        const config = options.generator.config;
        var _config_clientOutput, _ref;
        const prismaOutputLocation = (_ref = (_config_clientOutput = config.clientOutput) !== null && _config_clientOutput !== void 0 ? _config_clientOutput : (_options_otherGenerators_find = options.otherGenerators.find((gen)=>gen.provider.value === 'prisma-client')) === null || _options_otherGenerators_find === void 0 ? void 0 : (_options_otherGenerators_find_output = _options_otherGenerators_find.output) === null || _options_otherGenerators_find_output === void 0 ? void 0 : _options_otherGenerators_find_output.value) !== null && _ref !== void 0 ? _ref : options.otherGenerators.find((gen)=>gen.provider.value === 'prisma-client-js').output.value;
        const usingPureTsGenerator = !!options.otherGenerators.find((gen)=>gen.provider.value === 'prisma-client');
        const extensions = [
            '.js',
            '.ts',
            '.cjs',
            '.mjs',
            '.cts',
            '.mts'
        ];
        const prismaLocation = prismaOutputLocation.startsWith('@') || extensions.some((ext)=>prismaOutputLocation.endsWith(ext)) ? prismaOutputLocation : prismaOutputLocation.startsWith('./') ? `./${_nodepath.posix.join(prismaOutputLocation, usingPureTsGenerator ? 'client.js' : 'index.js')}` : _nodepath.posix.join(prismaOutputLocation, usingPureTsGenerator ? 'client.js' : 'index.js');
        if (!prismaLocation) {
            throw new Error('Unable to find prisma client output when generating pothos types');
        }
        var _options_generator_output_value;
        const outputLocation = (_options_generator_output_value = (_options_generator_output = options.generator.output) === null || _options_generator_output === void 0 ? void 0 : _options_generator_output.value) !== null && _options_generator_output_value !== void 0 ? _options_generator_output_value : defaultOutput;
        const prismaTypes = buildTypes(options.dmmf, config);
        await generateOutput(options.dmmf, prismaTypes, prismaLocation, outputLocation, config, outputLocation === defaultOutput ? 'cjs' : undefined);
        if (outputLocation === defaultOutput) {
            await generateOutput(options.dmmf, prismaTypes, prismaLocation, (0, _nodepath.join)(outputLocation, '../esm/generated.d.ts'), config, 'esm');
        }
    }
});
function trimDmmf(dmmf, documentation = false) {
    const trimmed = {
        datamodel: {
            models: {}
        }
    };
    for (const model of dmmf.datamodel.models){
        trimmed.datamodel.models[model.name] = {
            fields: model.fields.map((field)=>({
                    type: field.type,
                    kind: field.kind,
                    name: field.name,
                    isRequired: field.isRequired,
                    isList: field.isList,
                    hasDefaultValue: field.hasDefaultValue,
                    isUnique: field.isUnique,
                    isId: field.isId,
                    relationName: field.relationName,
                    relationFromFields: field.relationFromFields,
                    isUpdatedAt: field.isUpdatedAt,
                    documentation: documentation ? field.documentation : undefined
                })),
            primaryKey: model.primaryKey ? {
                name: model.primaryKey.name,
                fields: model.primaryKey.fields
            } : null,
            uniqueIndexes: model.uniqueIndexes.map((index)=>({
                    name: index.name,
                    fields: index.fields
                })),
            documentation: documentation ? model.documentation : undefined
        };
    }
    return trimmed;
}
// biome-ignore lint/suspicious/useAwait: needs to be async
async function generateOutput(dmmf, prismaTypes, prismaLocation, outputLocation, config, datamodel) {
    const prismaImportStatement = _typescript.default.factory.createImportDeclaration(...modifiersArg, [], _typescript.default.factory.createImportClause(true, undefined, _typescript.default.factory.createNamedImports([
        _typescript.default.factory.createImportSpecifier(false, undefined, _typescript.default.factory.createIdentifier('Prisma')),
        ...dmmf.datamodel.models.map((model)=>_typescript.default.factory.createImportSpecifier(false, undefined, _typescript.default.factory.createIdentifier(model.name)))
    ])), _typescript.default.factory.createStringLiteral(prismaLocation));
    var _config_pluginPath;
    const dmmfImportStatement = _typescript.default.factory.createImportDeclaration(...modifiersArg, [], _typescript.default.factory.createImportClause(true, undefined, _typescript.default.factory.createNamedImports([
        _typescript.default.factory.createImportSpecifier(false, undefined, _typescript.default.factory.createIdentifier('PothosPrismaDatamodel'))
    ])), _typescript.default.factory.createStringLiteral((_config_pluginPath = config.pluginPath) !== null && _config_pluginPath !== void 0 ? _config_pluginPath : '@pothos/plugin-prisma'));
    const printer = _typescript.default.createPrinter({});
    const sourcefile = _typescript.default.createSourceFile(outputLocation, '', _typescript.ScriptTarget.ESNext, false, _typescript.ScriptKind.TS);
    const trimmedDatamodel = trimDmmf(dmmf, config.documentation === 'true');
    const dmmfExport = _typescript.default.factory.createFunctionDeclaration([
        _typescript.default.factory.createModifier(_typescript.SyntaxKind.ExportKeyword)
    ], undefined, 'getDatamodel', [], [], _typescript.default.factory.createTypeReferenceNode('PothosPrismaDatamodel'), outputLocation.endsWith('.d.ts') ? undefined : _typescript.default.factory.createBlock([
        _typescript.default.factory.createReturnStatement(_typescript.default.factory.createCallExpression(_typescript.default.factory.createPropertyAccessExpression(_typescript.default.factory.createIdentifier('JSON'), 'parse'), [], [
            _typescript.default.factory.createStringLiteral(JSON.stringify(trimmedDatamodel))
        ]))
    ]));
    const nodes = _typescript.default.factory.createNodeArray(config.generateDatamodel !== 'false' ? [
        prismaImportStatement,
        dmmfImportStatement,
        prismaTypes,
        dmmfExport
    ] : [
        prismaImportStatement,
        prismaTypes
    ]);
    const result = printer.printList(_typescript.ListFormat.SourceFileStatements, nodes, sourcefile);
    (0, _nodefs.mkdirSync)((0, _nodepath.dirname)(sourcefile.fileName), {
        recursive: true
    });
    (0, _nodefs.writeFileSync)(sourcefile.fileName, `/* eslint-disable */\n${result}`);
    if (outputLocation.endsWith('.d.ts')) {
        if (config.generateDatamodel === 'true' && datamodel === 'cjs') {
            (0, _nodefs.writeFileSync)(outputLocation.replace(/\.d\.ts$/, '.js'), `/* eslint-disable */\nmodule.exports = { getDatamodel: () => JSON.parse(${JSON.stringify(JSON.stringify(trimmedDatamodel))}) }\n`);
        }
        if (config.generateDatamodel === 'true' && datamodel === 'esm') {
            (0, _nodefs.writeFileSync)(outputLocation.replace(/\.d\.ts$/, '.js'), `/* eslint-disable */\nexport function getDatamodel() { return JSON.parse(${JSON.stringify(JSON.stringify(trimmedDatamodel))}) }\n`);
        }
    }
}
function buildTypes(dmmf, config) {
    function getOrderByTypeName(type) {
        var _dmmf_schema_inputObjectTypes_prisma;
        const possibleTypes = [
            `${type}OrderByWithRelationInput`,
            `${type}OrderByWithRelationAndSearchRelevanceInput`
        ];
        const orderBy = (_dmmf_schema_inputObjectTypes_prisma = dmmf.schema.inputObjectTypes.prisma) === null || _dmmf_schema_inputObjectTypes_prisma === void 0 ? void 0 : _dmmf_schema_inputObjectTypes_prisma.find((inputType)=>possibleTypes.includes(inputType.name));
        if (!orderBy) {
            return possibleTypes[0];
        }
        return orderBy.name;
    }
    const prismaUtils = config.prismaUtils === 'true';
    const modelTypes = dmmf.datamodel.models.map((model)=>{
        var _dmmf_schema_inputObjectTypes_prisma;
        const relations = model.fields.filter((field)=>!!field.relationName);
        const listRelations = model.fields.filter((field)=>!!field.relationName && field.isList);
        const createInputUnavailable = !((_dmmf_schema_inputObjectTypes_prisma = dmmf.schema.inputObjectTypes.prisma) === null || _dmmf_schema_inputObjectTypes_prisma === void 0 ? void 0 : _dmmf_schema_inputObjectTypes_prisma.some((input)=>input.name === `${model.name}CreateInput`));
        return _typescript.default.factory.createPropertySignature([], model.name, undefined, _typescript.default.factory.createTypeLiteralNode([
            _typescript.default.factory.createPropertySignature([], 'Name', undefined, _typescript.default.factory.createLiteralTypeNode(_typescript.default.factory.createStringLiteral(model.name))),
            _typescript.default.factory.createPropertySignature([], 'Shape', undefined, _typescript.default.factory.createTypeReferenceNode(model.name)),
            _typescript.default.factory.createPropertySignature([], 'Include', undefined, relations.length > 0 ? _typescript.default.factory.createTypeReferenceNode(`Prisma.${model.name}Include`) : _typescript.default.factory.createTypeReferenceNode('never')),
            _typescript.default.factory.createPropertySignature([], 'Select', undefined, _typescript.default.factory.createTypeReferenceNode(`Prisma.${model.name}Select`)),
            _typescript.default.factory.createPropertySignature([], 'OrderBy', undefined, _typescript.default.factory.createTypeReferenceNode(`Prisma.${getOrderByTypeName(model.name)}`)),
            _typescript.default.factory.createPropertySignature([], 'WhereUnique', undefined, _typescript.default.factory.createTypeReferenceNode(`Prisma.${model.name}WhereUniqueInput`)),
            _typescript.default.factory.createPropertySignature([], 'Where', undefined, _typescript.default.factory.createTypeReferenceNode(`Prisma.${model.name}WhereInput`)),
            ...prismaUtils ? [
                _typescript.default.factory.createPropertySignature([], 'Create', undefined, createInputUnavailable ? _typescript.default.factory.createTypeLiteralNode([]) : _typescript.default.factory.createTypeReferenceNode(`Prisma.${model.name}CreateInput`)),
                _typescript.default.factory.createPropertySignature([], 'Update', undefined, _typescript.default.factory.createTypeReferenceNode(`Prisma.${model.name}UpdateInput`))
            ] : [
                _typescript.default.factory.createPropertySignature([], 'Create', undefined, _typescript.default.factory.createTypeLiteralNode([])),
                _typescript.default.factory.createPropertySignature([], 'Update', undefined, _typescript.default.factory.createTypeLiteralNode([]))
            ],
            _typescript.default.factory.createPropertySignature([], 'RelationName', undefined, relations.length > 0 ? _typescript.default.factory.createUnionTypeNode(relations.map((field)=>_typescript.default.factory.createLiteralTypeNode(_typescript.default.factory.createStringLiteral(field.name)))) : _typescript.default.factory.createTypeReferenceNode('never')),
            _typescript.default.factory.createPropertySignature([], 'ListRelations', undefined, listRelations.length > 0 ? _typescript.default.factory.createUnionTypeNode(listRelations.map((field)=>_typescript.default.factory.createLiteralTypeNode(_typescript.default.factory.createStringLiteral(field.name)))) : _typescript.default.factory.createTypeReferenceNode('never')),
            _typescript.default.factory.createPropertySignature([], 'Relations', undefined, _typescript.default.factory.createTypeLiteralNode(relations.map((field)=>{
                const typeName = field.type;
                return _typescript.default.factory.createPropertySignature([], field.name, undefined, _typescript.default.factory.createTypeLiteralNode([
                    _typescript.default.factory.createPropertySignature([], 'Shape', undefined, field.isList ? _typescript.default.factory.createArrayTypeNode(_typescript.default.factory.createTypeReferenceNode(typeName)) : field.isRequired ? _typescript.default.factory.createTypeReferenceNode(typeName) : _typescript.default.factory.createUnionTypeNode([
                        _typescript.default.factory.createTypeReferenceNode(typeName),
                        _typescript.default.factory.createLiteralTypeNode(_typescript.default.factory.createNull())
                    ])),
                    _typescript.default.factory.createPropertySignature([], 'Name', undefined, _typescript.default.factory.createLiteralTypeNode(_typescript.default.factory.createStringLiteral(typeName))),
                    _typescript.default.factory.createPropertySignature([], 'Nullable', undefined, _typescript.default.factory.createLiteralTypeNode(field.isRequired ? _typescript.default.factory.createFalse() : _typescript.default.factory.createTrue()))
                ]));
            })))
        ]));
    });
    return _typescript.default.factory.createInterfaceDeclaration(...modifiersArg, [
        _typescript.default.factory.createModifier(_typescript.SyntaxKind.ExportKeyword),
        _typescript.default.factory.createModifier(_typescript.SyntaxKind.DefaultKeyword)
    ], 'PrismaTypes', [], [], modelTypes);
}
