// @pothos/plugin-relay/esm/index.js
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    }
    else {
        obj[key] = value;
    }
    return obj;
}
import './global-types.js';
import './field-builder.js';
import './input-field-builder.js';
import './schema-builder.js';
import SchemaBuilder, { BasePlugin, createInputValueMapper, mapInputFields } from '@pothos/core';
import { internalDecodeGlobalID } from './utils/internal.js';
export * from './node-ref.js';
export * from './types.js';
export * from './utils/index.js';
const pluginName = "relay";
export default pluginName;
export class PothosRelayPlugin extends BasePlugin {
    onOutputFieldConfig(fieldConfig) {
        const argMappings = mapInputFields(fieldConfig.args, this.buildCache, (inputField) => {
            var _inputField_extensions;
            if ((_inputField_extensions = inputField.extensions) === null || _inputField_extensions === void 0 ? void 0 : _inputField_extensions.isRelayGlobalID) {
                var _inputField_extensions1, _inputField_extensions2;
                var _inputField_extensions_relayGlobalIDFor, _ref;
                return (_ref = (_inputField_extensions_relayGlobalIDFor = (_inputField_extensions1 = inputField.extensions) === null || _inputField_extensions1 === void 0 ? void 0 : _inputField_extensions1.relayGlobalIDFor) !== null && _inputField_extensions_relayGlobalIDFor !== void 0 ? _inputField_extensions_relayGlobalIDFor : (_inputField_extensions2 = inputField.extensions) === null || _inputField_extensions2 === void 0 ? void 0 : _inputField_extensions2.relayGlobalIDAlwaysParse) !== null && _ref !== void 0 ? _ref : false;
            }
            return null;
        }, this.mappingCache);
        if (!argMappings) {
            return fieldConfig;
        }
        const argMapper = createInputValueMapper(argMappings, (globalID, mappings, ctx, info) => {
            var _mappings_value;
            return internalDecodeGlobalID(this.builder, String(globalID), ctx, info, (_mappings_value = mappings.value) !== null && _mappings_value !== void 0 ? _mappings_value : false);
        });
        var _fieldConfig_argMappers;
        return {
            ...fieldConfig,
            argMappers: [
                ...(_fieldConfig_argMappers = fieldConfig.argMappers) !== null && _fieldConfig_argMappers !== void 0 ? _fieldConfig_argMappers : [],
                (args, context, info) => argMapper(args, undefined, context, info)
            ]
        };
    }
    constructor(...args) {
        super(...args), _define_property(this, "mappingCache", new Map());
    }
}
SchemaBuilder.registerPlugin(pluginName, PothosRelayPlugin, {
    v3: (options) => {
        var _options_relayOptions, _options_relayOptions1, _options_relayOptions2, _options_relayOptions3, _options_relayOptions4, _options_relayOptions5, _options_relayOptions_edgesFieldOptions, _options_relayOptions_nodeFieldOptions;
        var _options_relayOptions_clientMutationId, _options_relayOptions_cursorType, _options_relayOptions_edgeCursorType, _ref, _options_relayOptions_pageInfoCursorType, _ref1, _options_relayOptions_edgesFieldOptions_nullable, _options_relayOptions_nodeFieldOptions_nullable, _options_relayOptions_brandLoadedObjects;
        return {
            relayOptions: undefined,
            relay: {
                ...options.relayOptions,
                clientMutationId: (_options_relayOptions_clientMutationId = (_options_relayOptions = options.relayOptions) === null || _options_relayOptions === void 0 ? void 0 : _options_relayOptions.clientMutationId) !== null && _options_relayOptions_clientMutationId !== void 0 ? _options_relayOptions_clientMutationId : "required",
                cursorType: (_options_relayOptions_cursorType = (_options_relayOptions1 = options.relayOptions) === null || _options_relayOptions1 === void 0 ? void 0 : _options_relayOptions1.cursorType) !== null && _options_relayOptions_cursorType !== void 0 ? _options_relayOptions_cursorType : "ID",
                edgeCursorType: (_ref = (_options_relayOptions_edgeCursorType = (_options_relayOptions2 = options.relayOptions) === null || _options_relayOptions2 === void 0 ? void 0 : _options_relayOptions2.edgeCursorType) !== null && _options_relayOptions_edgeCursorType !== void 0 ? _options_relayOptions_edgeCursorType : (_options_relayOptions3 = options.relayOptions) === null || _options_relayOptions3 === void 0 ? void 0 : _options_relayOptions3.cursorType) !== null && _ref !== void 0 ? _ref : "String",
                pageInfoCursorType: (_ref1 = (_options_relayOptions_pageInfoCursorType = (_options_relayOptions4 = options.relayOptions) === null || _options_relayOptions4 === void 0 ? void 0 : _options_relayOptions4.pageInfoCursorType) !== null && _options_relayOptions_pageInfoCursorType !== void 0 ? _options_relayOptions_pageInfoCursorType : (_options_relayOptions5 = options.relayOptions) === null || _options_relayOptions5 === void 0 ? void 0 : _options_relayOptions5.cursorType) !== null && _ref1 !== void 0 ? _ref1 : "String",
                edgesFieldOptions: {
                    ...options.relayOptions.edgesFieldOptions,
                    nullable: (_options_relayOptions_edgesFieldOptions_nullable = (_options_relayOptions_edgesFieldOptions = options.relayOptions.edgesFieldOptions) === null || _options_relayOptions_edgesFieldOptions === void 0 ? void 0 : _options_relayOptions_edgesFieldOptions.nullable) !== null && _options_relayOptions_edgesFieldOptions_nullable !== void 0 ? _options_relayOptions_edgesFieldOptions_nullable : {
                        list: false,
                        items: true
                    }
                },
                nodeFieldOptions: {
                    ...options.relayOptions.nodeFieldOptions,
                    nullable: (_options_relayOptions_nodeFieldOptions_nullable = (_options_relayOptions_nodeFieldOptions = options.relayOptions.nodeFieldOptions) === null || _options_relayOptions_nodeFieldOptions === void 0 ? void 0 : _options_relayOptions_nodeFieldOptions.nullable) !== null && _options_relayOptions_nodeFieldOptions_nullable !== void 0 ? _options_relayOptions_nodeFieldOptions_nullable : false
                },
                brandLoadedObjects: (_options_relayOptions_brandLoadedObjects = options.relayOptions.brandLoadedObjects) !== null && _options_relayOptions_brandLoadedObjects !== void 0 ? _options_relayOptions_brandLoadedObjects : false
            }
        };
    }
});
