// @pothos/core/lib/types/type-params.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get abstractReturnShapeKey () {
        return abstractReturnShapeKey;
    },
    get inputFieldShapeKey () {
        return inputFieldShapeKey;
    },
    get inputShapeKey () {
        return inputShapeKey;
    },
    get outputFieldShapeKey () {
        return outputFieldShapeKey;
    },
    get outputShapeKey () {
        return outputShapeKey;
    },
    get parentShapeKey () {
        return parentShapeKey;
    },
    get typeBrandKey () {
        return typeBrandKey;
    }
});
const outputShapeKey = Symbol.for('Pothos.outputShapeKey');
const parentShapeKey = Symbol.for('Pothos.parentShapeKey');
const abstractReturnShapeKey = Symbol.for('Pothos.abstractReturnShapeKey');
const inputShapeKey = Symbol.for('Pothos.inputShapeKey');
const inputFieldShapeKey = Symbol.for('Pothos.inputFieldShapeKey');
const outputFieldShapeKey = Symbol.for('Pothos.outputFieldShapeKey');
const typeBrandKey = Symbol.for('Pothos.typeBrandKey');
